src/web/components/Onwards/Onwards.tsx

```diff
@@ -1,28 +1,170 @@
-    pathId: OnwardsIdType;
-    component: React.ElementType;
+    isPaidContent: boolean;
+    showRelatedContent: boolean;
+    keywordIds: string[];
+    contentType: string;
 };
 
-export const Onwards = ({ ajaxUrl, pageId, pathId, component }: Props) => {
-    const url = `${new URL(`${pathId}/${pageId}`, ajaxUrl)}.json`;
-    const { data, error } = useApi<OnwardsType>(url);
+export const Onwards = ({
+    ajaxUrl,
+    hasRelated,
+    hasStoryPackage,
+    isAdFreeUser,
+    pageId,
+    isPaidContent,
+    showRelatedContent,
+    keywordIds,
+    contentType,
+}: Props) => {
+    const onwardSections: OnwardsType[] = [];
 
-    if (error) {
-        window.guardian.modules.sentry.reportError(error, 'story-package');
-        return null;
+    if (hasStoryPackage) {
+        // Always fetch the story package if it exists
+        const { data } = useApi(
+            joinUrl([ajaxUrl, 'story-package', `${pageId}.json?dcr=true`]),
+        );
+
+        const storyPackage = data;
+
+        if (data && data.trails && data.trails.length > 7) {
+            onwardSections.push({
+                heading: storyPackage.heading,
+                trails: storyPackage.trails,
+                layout: 'fourAndFour',
+            });
+        }
+    } else if (isAdFreeUser && isPaidContent) {
+        // Don't show any related content (other than story packages) for
+        // adfree users when the content is paid for
+    } else if (showRelatedContent && hasRelated) {
+        // Related content can be a collection of articles based on
+        // two things, 1: A popular tag, or 2: A generic text match
+        const popularTag = firstPopularTag(keywordIds, isPaidContent);
+
+        let relatedUrl;
+        if (popularTag) {
+            // Use popular in tag endpoint
+            relatedUrl = `/popular-in-tag/${popularTag}.json`;
+        } else {
+            // Default to generic related endpoint
+            relatedUrl = `/related/${pageId}.json`;
+        }
+        relatedUrl += '?dcr=true';
+
+        // --- Tag excludes --- //
+        const tagsToExclude = [];
+        // Exclude ad features from non-ad feature content
+        if (!isPaidContent) {
+            tagsToExclude.push('tone/advertisement-features');
+        }
+        // We don't want to show professional network content on videos or interactives
+        if (
+            contentType.toLowerCase() === 'video' ||
+            contentType.toLowerCase() === 'interactive'
+        ) {
+            tagsToExclude.push('guardian-professional/guardian-professional');
+        }
+
+        // Add any exclude tags to the url
+        if (tagsToExclude.length > 0) {
+            const queryParams = tagsToExclude.map(tag => `exclude-tag=${tag}`);
+            relatedUrl += `&${queryParams.join('&')}`;
+        }
+
+        const { data } = useApi(joinUrl([ajaxUrl, relatedUrl]));
+
+        const relatedContent = data;
+
+        if (data && data.trails && data.trails.length > 7) {
```

Comment:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining maybe?

Reply:
Is that working with our babel/lint setup now? That'd be 🕺

Reply:
It is working. It is🕺